synapse/module_api/__init__.py

```diff
@@ -174,6 +174,34 @@ def email_app_name(self) -> str:
         """The application name configured in the homeserver's configuration."""
         return self._hs.config.email.email_app_name
 
+    async def get_user_by_id(self, user_id: str) -> Optional[User]:
+        """Get User by user_id
+
+        Args:
+            user_id: Fully qualified user id.
+        Returns:
+            User object if a user was found, otherwise None
+        """
+        user_data = await self._store.get_user_by_id(user_id)
+        if user_data:
+            app_service = None
+            if user_data.get("appservice_id"):
+                app_service = self._store.get_app_service_by_id(
+                    user_data.get("appservice_id")
+                )
+            return User(
+                user=UserID.from_string(user_data.get("name")),
+                creation_ts=user_data.get("creation_ts"),
+                consent_version=user_data.get("consent_version"),
+                consent_server_notice_sent=user_data.get("consent_server_notice_sent"),
+                app_service=app_service,
+                shadow_banned=user_data.get("shadow_banned"),
+                deactivated=user_data.get("deactivated"),
+                is_guest=user_data.get("is_guest"),
+                admin=user_data.get("admin"),
+                user_type=user_data.get("user_type"),
+            )
```

Comment:
I think this logic belongs outside the module_api, which is meant to be a very thin wrapper.

Maybe better to add a `get_user_info_by_id` alongside `store.get_user_by_id`, and call straight through to that here.

(We can also then mark `_store.get_user_by_id` deprecated, since its interface is horrible.)

Reply:
I believe this could be up for a new review. The last commit is due to mypy, it feels wrong needing to have that but I'm not sure if there is another way around things?

Probably best deprecate `_store.get_user_by_id` in a follow-up pr, I'm happy to do that?